pkg/operator/ceph/object/objectstore.go

```diff
@@ -661,44 +672,52 @@ func CreatePools(context *Context, clusterSpec *cephv1.ClusterSpec, metadataPool
 	return nil
 }
 
-func createSimilarPools(context *Context, pools []string, clusterSpec *cephv1.ClusterSpec, poolSpec cephv1.PoolSpec, pgCount, ecProfileName string) error {
+func createSimilarPools(ctx *Context, pools []string, clusterSpec *cephv1.ClusterSpec, poolSpec cephv1.PoolSpec, pgCount, ecProfileName string) error {
+	g, _ := errgroup.WithContext(context.TODO())
+
 	for _, pool := range pools {
-		// create the pool if it doesn't exist yet
-		name := poolName(context.Name, pool)
-		if poolDetails, err := ceph.GetPoolDetails(context.Context, context.clusterInfo, name); err != nil {
-			// If the ceph config has an EC profile, an EC pool must be created. Otherwise, it's necessary
-			// to create a replicated pool.
-			var err error
-			if poolSpec.IsErasureCoded() {
-				// An EC pool backing an object store does not need to enable EC overwrites, so the pool is
-				// created with that property disabled to avoid unnecessary performance impact.
-				err = ceph.CreateECPoolForApp(context.Context, context.clusterInfo, name, ecProfileName, poolSpec, pgCount, AppName, false /* enableECOverwrite */)
+		// Avoid the loop re-using the same value with a closure
+		pool := pool
+		g.Go(func() error {
+			// create the pool if it doesn't exist yet
+			name := poolName(ctx.Name, pool)
+			if poolDetails, err := ceph.GetPoolDetails(ctx.Context, ctx.clusterInfo, name); err != nil {
+				// If the ceph config has an EC profile, an EC pool must be created. Otherwise, it's necessary
+				// to create a replicated pool.
+				var err error
+				if poolSpec.IsErasureCoded() {
+					// An EC pool backing an object store does not need to enable EC overwrites, so the pool is
+					// created with that property disabled to avoid unnecessary performance impact.
+					err = ceph.CreateECPoolForApp(ctx.Context, ctx.clusterInfo, name, ecProfileName, poolSpec, pgCount, AppName, false /* enableECOverwrite */)
+				} else {
+					err = ceph.CreateReplicatedPoolForApp(ctx.Context, ctx.clusterInfo, clusterSpec, name, poolSpec, pgCount, AppName)
+				}
+				if err != nil {
+					return errors.Wrapf(err, "failed to create pool %s for object store %s.", name, ctx.Name)
```

Comment:
How does the errgroup behave when there is a failure? Will it retry or log the error?

Reply:
The error looks like this:

```
2021-06-09 08:26:05.673575 I | ceph-spec: adding finalizer "cephobjectstore.ceph.rook.io" on "my-store"
2021-06-09 08:26:05.681799 E | ceph-object-controller: failed to set object store "rook-ceph/my-store" status to "Progressing". failed to update object "my-store" status: Operation cannot be fulfilled on cephobjectstores.ceph.rook.io "my-store": the object has been modified; please apply your changes to the latest version and try again
2021-06-09 08:26:05.686295 I | op-mon: parsing mon endpoints: a=10.100.223.211:6789,b=10.111.63.108:6789,c=10.108.123.222:6789
2021-06-09 08:26:05.998193 I | ceph-object-controller: reconciling object store deployments
2021-06-09 08:26:06.008034 I | ceph-object-controller: ceph object store gateway service running at 10.107.192.238
2021-06-09 08:26:06.008053 I | ceph-object-controller: reconciling object store pools
2021-06-09 08:26:06.640821 I | ceph-object-controller: pool size is changed from 3 to 1
2021-06-09 08:26:07.110102 I | cephclient: setting pool property "pg_num_min" to "8" on pool "my-store.rgw.meta"
2021-06-09 08:26:08.431578 E | ceph-object-controller: failed to set object store "rook-ceph/my-store" status to "Failure". failed to update object "my-store" status: Operation cannot be fulfilled on cephobjectstores.ceph.rook.io "my-store": the object has been modified; please apply your changes to the latest version and try again
2021-06-09 08:26:08.431671 E | ceph-object-controller: failed to reconcile failed to create object store deployments: failed to create object pools: failed to create metadata pools: failed to create pool my-store.rgw.log for object store my-store.: failed to create replicated pool my-store.rgw.log. no valid command found; 10 closest matches:
osd pool stats [<pool_name>]
osd pool scrub <who>...
osd pool deep-scrub <who>...
osd pool repair <who>...
osd pool force-recovery <who>...
osd pool force-backfill <who>...
osd pool cancel-force-recovery <who>...
osd pool cancel-force-backfill <who>...
osd pool autoscale-status
osd pool mksnap <pool> <snap>
Traceback (most recent call last):
  File "/usr/bin/ceph", line 1310, in <module>
    retval = main()
  File "/usr/bin/ceph", line 1256, in main
    outf.write(outbuf)
TypeError: a bytes-like object is required, not 'str'
. : no valid command found; 10 closest matches:
osd pool stats [<pool_name>]
osd pool scrub <who>...
osd pool deep-scrub <who>...
osd pool repair <who>...
osd pool force-recovery <who>...
osd pool force-backfill <who>...
osd pool cancel-force-recovery <who>...
osd pool cancel-force-backfill <who>...
osd pool autoscale-status
osd pool mksnap <pool> <snap>
Traceback (most recent call last):
  File "/usr/bin/ceph", line 1310, in <module>
    retval = main()
  File "/usr/bin/ceph", line 1256, in main
    outf.write(outbuf)
TypeError: a bytes-like object is required, not 'str'
. 
```

So returns the error and calls reconcile again.

Reply:
Ok so the waitgroup finishes the same whether it's successful or fails, then the reconcile restarts with the same pattern we already had.